web/src/settings_components.ts

```diff
@@ -359,6 +359,71 @@ function get_message_retention_setting_value(
     return check_valid_number_input(custom_input_val);
 }
 
+type FieldData = Record<string, Record<string, string> | string>;
+
+function read_select_field_data_from_form(
+    $profile_field_form: JQuery,
+    old_field_data: FieldData,
+): FieldData {
+    const field_data: FieldData = {};
+    let field_order = 1;
+
+    const old_option_value_map = new Map();
+    if (old_field_data !== undefined) {
+        for (const [value, choice] of Object.entries(old_field_data)) {
+            assert(typeof choice !== "string");
+            old_option_value_map.set(choice.text, value);
+        }
+    }
+    $profile_field_form.find("div.choice-row").each(function (this: HTMLElement) {
+        const text = $(this).find("input")[0].value;
+        if (text) {
+            if (old_option_value_map.get(text) !== undefined) {
+                // Resetting the data-value in the form is
+                // important if the user removed an option string
+                // and then added it back again before saving
+                // changes.
```

Comment:
Again, this comment is pretty good, but even better would be to provide more context to the reader about the problem statement. Something like

> In this case, the user is adding a new choice with the same text as an existing choice before saving their changes. Make sure that we restore the original `data-value` for this choice, so that ... .

Though I'm a bit confused -- what happens to whatever `data-value` had been generated originally for the option row -- I guess it'll just be an unused integer?

Reply:
Yes, it will be just an unused integer. The deleted rows in some cases will also result in their value integer remaining unused, e.g. if the values are `[1, 2, 3, 4, 5]` and we delete `2`, it will remain unused, even if we add no new elements. Not sure if we should do anything about it or not, since the value is only used internally and not user facing. 

While looking into this, I noticed a bug in this PR, opened an issue at: #29879. I'll also try to fix the other comments in the PR for that issue in a different refactor commit